File: app/src/main/java/com/example/data/repository/FavoriteRepository.kt
package com.example.data.repository

import android.content.Context
import com.example.data.service.PrefsHelper
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

interface FavoriteRepository {
    val wishlist: StateFlow<Set<String>>
    fun toggleWishlist(hotelId: String)
}

class FavoriteRepositoryImpl(context: Context) : FavoriteRepository {
    private val prefsHelper = PrefsHelper(context)

    companion object {
        private val _wishlist = MutableStateFlow<Set<String>>(emptySet())
        val wishlistFlow = _wishlist.asStateFlow()
        private var isInitialized = false
    }

    override val wishlist: StateFlow<Set<String>> = wishlistFlow

    init {
        synchronized(this) {
            if (!isInitialized) {
                val wishlistSet = prefsHelper.prefs.getStringSet("wishlist_hotel_ids", emptySet()) ?: emptySet()
                _wishlist.value = wishlistSet
                isInitialized = true
            }
        }
    }

    override fun toggleWishlist(hotelId: String) {
        val updated = _wishlist.value.toMutableSet()
        if (updated.contains(hotelId)) {
            updated.remove(hotelId)
        } else {
            updated.add(hotelId)
        }
        _wishlist.value = updated
        prefsHelper.prefs.edit().putStringSet("wishlist_hotel_ids", updated).apply()
    }
}

File: lib/repositories/favorite_repository.dart
import 'package:flutter/foundation.dart';
import '../service/prefs_helper.dart';

class FavoriteRepository {
  static final wishlist = ValueNotifier<Set<String>>({});
  static bool _isInitialized = false;

  static void init() {
    if (!_isInitialized) {
      loadWishlist();
      _isInitialized = true;
    }
  }

  static void loadWishlist() {
    final list = PrefsHelper.prefs.getStringList("wishlist_hotel_ids") ?? [];
    wishlist.value = list.toSet();
  }

  static void saveWishlist() {
    PrefsHelper.prefs.setStringList("wishlist_hotel_ids", wishlist.value.toList());
  }

  static void toggleWishlist(String hotelId) {
    final updated = Set<String>.from(wishlist.value);
    if (updated.contains(hotelId)) {
      updated.remove(hotelId);
    } else {
      updated.add(hotelId);
    }
    wishlist.value = updated;
    saveWishlist();
  }
}
